6 Loading Related Data

deakgeek opened the pull request:
#6

deakgeek committed:
Added code snippets for loading related data tutorial

deakgeek committed:
Added code snippet for the field not included section

arynhuck committed:
Loading related data content

Added the content for loading related data. This includes the tutorial
itself, adding the navigation stuff for TOC, and the new reference
pages.

deakgeek committed:
Updated code snippets with new tutorial changes

deakgeek committed:
Merge branch '6-loading_related_data' of https://github.com/LadyNaggaga/training-tutorials into 6-loading_related_data

deakgeek committed:
Updated code snippets to a list of Editions in Book.cs and a list of Books in Author.cs

deakgeek committed:
Organized Code Snippets

arynhuck committed:
merge master -> 6-loading_related_data

This is the merge commit before we have this branch merged into master.

arynhuck committed:
Merge branch '6-loading_related_data' of https://github.com/LadyNaggaga/training-tutorials into 6-loading_related_data

arynhuck commented:
Merged master in. should be good to :shipit: soon.

deakgeek committed:
Moved Installing EF Core to the correct tutorial

arynhuck committed:
Words to letters in LINQ queries

This is to appease Ryan, but also cuz like conventions. Went through and
changed everywhere it used a word (which also happened to be the real
variable's name) to a litter to use in the LINQ query.
This was done in both the tutorial content file and the code snippets
files.

arynhuck committed:
Making it LibraryContext consistently

Accidentally forgot to change little bits where we had it as book or
author context. Have now changed it.

arynhuck committed:
Merge master -> branch

Preparation for PR. Merge commit.

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-and-then-include-method-chain/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
`Publisher.cs` also needs to exist for this tutorial

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method-chain/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
`Publisher.cs` needs to exist for this tutorial

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
`Publisher.cs` needs to exist for this tutorial as well

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/shared/Book.cs:
```diff
@@ -1,9 +1,9 @@
 public class Book
 {
-    public int BookId { get; set; }
+    public int Id { get; set; }
```
Should be `BookId`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/shared/Edition.cs:
```diff
@@ -0,0 +1,7 @@
+public class Edition
+{
+    public int EditionId { get; set; }
+    public int BookId { get; set; }
```
This should be `public Book Book { get; set; }`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/shared/Edition.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public int EditionId { get; set; }
+    public int BookId { get; set; }
+    public int PublisherId { get; set; }
```
This should be `public Publisher Publisher { get; set; }`

arynhuck replied:
🌮

rdahlman3 commented:
Will need to edit `LibraryContext.cs` to include a `DbSet` for both `Editions` and `Publishers`

rdahlman3 commented on content/efcore/getting-started/samples/shared/Book.cs:
```diff
@@ -1,9 +1,9 @@
 public class Book
```
Need to have `using System.Collections.Generic;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/shared/Author.cs:
```diff
@@ -3,4 +3,5 @@ public class Author
     public int AuthorId { get; set; }
```
At the top of this file need to have `using System.Collections.Generic;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-and-then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Any of the files that use `Include` and/or `ThenInclude` will need `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Will need `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Will need `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/count-method/Program.cs:
```diff
@@ -0,0 +1,21 @@
+                .Collection(a => a.Books)
+                .Query()
+                .Count();
+            Console.WriteLine("Count: {0}", author.Books.Count);
```
Instead of printing `author.Books.Count`, should be printing `numBooks`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/count-method/Program.cs:
```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Collections.Generic;
```
Unnecessary

arynhuck replied:
Both or one?

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/field-not-included/Program.cs:
```diff
@@ -0,0 +1,16 @@
+using System;
+using System.Collections.Generic;
```
Unnecessary

arynhuck replied:
both or one?

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-and-then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        using (var context = new LibraryContext())
+        {
+            var book = context.Books
```
Actually, I ended up changing it to the following code when testing because `Call of the Wild` only has one edition:
```
var book = context.Books
                .Include(b => b.Editions)
                .ThenInclude(e => e.Publisher)
                .Include(b => b.Author)
                .Single(b => b.Title == "The Scarlet Plague");

            Console.WriteLine(String.Format("{0}", book.Title));
            Console.WriteLine("Editions:");
            foreach (Edition edition in book.Editions) {
                Console.WriteLine(String.Format("{0}, {1}", edition.Publisher.Name, edition.Year));
            }
```

rdahlman3 replied:
I'm open to other options, but this is just something that I got to work well.

arynhuck replied:
does it matter that _Call of the Wild_ only has one edition?

rdahlman3 replied:
I'm fine with using it even though it only has one edition - however, we either need to change it so that it is `== "The Call of the Wild"` or `.Contains("Call of the Wild")`, because right now it is failing without the `The`

arynhuck replied:
I'll change it to contains

arynhuck replied:
🌮

arynhuck replied:
also changed so single is last

rdahlman3 commented on content/efcore/getting-started/loading-related-data.md:
```diff
@@ -0,0 +1,150 @@
+ 
+### Aggregate Functions 
+ 
+When using explicit loading, we can also use aggregate functions, such as count, max, min, and sum, which allow us to perform calculations on the data without loading it all into memory. The following example counts how many books in the library are by Willa Cather without having to load all of the books into memory: 
```
Should we have a note that this will only work if you have EF Core 1.1.0, like in the docs?

arynhuck replied:
Does:

> Note {.note}
> Aggregate functions are only available in EF Core versions 1.1.0 and newer.

work?

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-and-then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Need to have `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-and-then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

arynhuck replied:
Both of these lines or just `Using System;`?

rdahlman3 replied:
Just `using System.Collections.Generic;`, sorry
Whenever I have unnecessary, it is just on the line that it is commented on

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-method/Program.cs:
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
```
Unnecessary

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-method/Program.cs:
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Add `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-method/Program.cs:
```diff
@@ -0,0 +1,17 @@
+        {
+            var book = context.Books
+                .Single(b => b.BookId == 1)
+                .Include(b => b.Authors);
```
`Include` needs to be right after `Books`, before `Single`

rdahlman3 replied:
Also, needs to be `b.Author`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-method/Program.cs:
```diff
@@ -0,0 +1,17 @@
+            var book = context.Books
+                .Single(b => b.BookId == 1)
+                .Include(b => b.Authors);
+            Console.WriteLine(String.Format("{0} - {1}", book.Title, book.Author));
```
To print out the actual information, need to add another `{2}` in the formatted string and change it from `book.Author` to `book.Author.FirstName, book.Author.LastName`

arynhuck replied:
Not sure how to do this

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/load-method/Program.cs:
```diff
@@ -0,0 +1,22 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Need `using Microsoft.EntityFrameworkCore;`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/load-method/Program.cs:
```diff
@@ -0,0 +1,22 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/load-method/Program.cs:
```diff
@@ -0,0 +1,22 @@
+                .Query()
+                .Where(b => b.Title.Contains("Huck"))
+                .Load();
+            Console.WriteLine(book.Title);
```
`book` doesn't exist yet...we could just do a foreach loop to show that it is the only book of his that was loaded, is there an example where we show it wasn't initially loaded? Cuz we could do that before loading, but it's also fine the way it is
 
```
foreach (Book book in author.Books)
 {
    Console.WriteLine(book.Title);
}
```

arynhuck replied:
We already show that things aren't initially loaded unless included in some expect at the very beginning of the tutorial

arynhuck replied:
🌮

arynhuck committed:
Minor code class changes

We had tiny things (naming, forgetting to include related classes) that
needed fixing!

arynhuck committed:
I don't know why this didn't commit with the others

Github is dumb or I am dumb (more likely)

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/min-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        using (var context = new LibraryContext())
+        {
+            var year = context.Entry(book)
```
Need to load a book before doing this, like in the `Count` example snippet - might be good to load a book with multiple editions like _The Scarlet Plague_

arynhuck replied:
This is like how the examples I could find do it...

rdahlman3 replied:
The code is all completely correct, except there's not actually an object in the `book` variable, so it has nothing to search on. A book from the db just needs to be fetched before trying to find the min year of its editions

arynhuck committed:
merge master -> branch

Getting this updated for hopeful PR :shipit: ?

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/min-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+                .Collection(b => b.Editions)
+                .Query()
+                .Min(e => e.Year);
+            Console.WriteLine("Minimum Year: {0}", book.Editions.Min);
```
Change to printing `year`

arynhuck replied:
🌮

arynhuck commented:
@rdahlman3 @helmoski changes have been made!
🌮

rdahlman3 reviewed (changes requested):
Started reviewing the code snippets to make sure they work

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/count-method/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
Need to add `Publisher.cs`

arynhuck replied:
why do all of these have READMEs? Why can't we just include a README in the directory above that says the classes that are included?

rdahlman3 replied:
I'm definitely fine with that, but don't know if that would involve changes to the work already done on the markdown engine, and we don't really have time to make those changes

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/field-not-included/README.md:
```diff
@@ -0,0 +1,9 @@
+- [Program.cs](Program.cs)
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
```
Need to add `Publisher.cs` and `Edition.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/include-method/README.md:
```diff
@@ -0,0 +1,9 @@
+- [Program.cs](Program.cs)
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
```
Need to add `Publisher.cs` and `Edition.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/load-method/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
Need to add `Publisher.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/min-method/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
Need to add `Publisher.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/Program.cs:
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Needs `using Microsoft.EntityFrameworkCore;`

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/Program.cs:
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/Program.cs:
```diff
@@ -0,0 +1,18 @@
+        {
+            var book = context.Books
+                .Single(b => b.BookId == 1)
+                .Include(b => b.Editions)
```
The `Include`s need to be before the `Single`

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/Program.cs:
```diff
@@ -0,0 +1,18 @@
+                .Single(b => b.BookId == 1)
+                .Include(b => b.Editions)
+                .Include(b => b.Author);
+            Console.WriteLine(String.Format("{0} - {1}", book.Author, book.Edition));
```
If we want to show the actual data, we will need to change it so that it prints `book.Author.FirstName` and `book.Author.LastName`. We also should do a foreach loop or something of the sort for editions, since it is a list.

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
Needs to include `Publisher.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/query-method/Program.cs:
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/query-method/Program.cs:
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Needs `using Microsoft.EntityFrameworkCore;`

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/query-method/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
+- [Edition.cs](../../shared/Edition.cs)
```
Needs `Publisher.cs`

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+        using (var context = new LibraryContext())
+        {
+            var author = context.Authors
+                .Single(a => a.LastName == "Douglass")
```
`Single` needs to be after all of the `Include` and `ThenInclude`s

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method-chain/Program.cs:
```diff
@@ -0,0 +1,19 @@
+                .Include(a => a.Books)
+                    .ThenInclude(b => b.Editions)
+                        .ThenInclude(e => e.Publisher);
+            Console.WriteLine(String.Format("{0} : {1} - {2}", book.Title, book.Editions, edition.Publisher));
```
This needs to be reworked, because the author is loaded, not the book, and both `Books` and `Editions` are `Lists`. Maybe we could just do `First` on the books, and then a `foreach` for the editions?

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
```
unnecessary

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+        using (var context = new LibraryContext())
+        {
+            var book = context.Books
+                .Single(b => b.BookId == 1)
```
`Single` needs to be after the `Include` and `ThenInclude`

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/then-include-method/Program.cs:
```diff
@@ -0,0 +1,18 @@
+                .Single(b => b.BookId == 1)
+                .Include(b => b.Editions)
+                    .ThenInclude(e => e.Publisher);
+            Console.WriteLine(String.Format("{0} - {1}", book.Editions, edition.Publisher));
```
Need a `foreach` loop to print out the editions and publishers

rdahlman3 commented on content/efcore/getting-started/samples/querying-data-tutorial/filtering-entities/README.md:
```diff
@@ -4,6 +4,6 @@ This code example demonstrates how to use EF Core to query a subset of entities
\ No newline at end of file
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
```
Needs `Publisher.cs` and `Editions.cs` as per discussion

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/querying-data-tutorial/loading-all-entities/README.md:
```diff
@@ -4,6 +4,6 @@ This code example demonstrates how to use EF Core to query all entities from the
\ No newline at end of file
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
```
Needs `Publisher.cs` and `Editions.cs` as per discussion

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/querying-data-tutorial/loading-single-entity/README.md:
```diff
@@ -4,6 +4,6 @@ This code example demonstrates how to use EF Core to query a single entity from
\ No newline at end of file
+- [LibraryContext.cs](../../shared/LibraryContext.cs)
+- [Book.cs](../../shared/Book.cs)
+- [Author.cs](../../shared/Author.cs)
```
Needs `Publisher.cs` and `Editions.cs` as per discussion

arynhuck replied:
🌮

rdahlman3 commented on content/efcore/getting-started/samples/shared/LibraryContext.cs:
```diff
@@ -6,6 +6,8 @@ public class LibraryContext : DbContext
 {
     public DbSet<Book> Books { get; set; }
     public DbSet<Author> Authors { get; set; }
+	public DbSet<Edition> Editions { get; set; }
```
I'm so sorry but the extra tab kills me internally

arynhuck replied:
There is no extra tab?
🌮

rdahlman3 replied:
Weird, GitHub must just be confused. My b

arynhuck committed:
More fixes

Someone else can work on this now.

deakgeek committed:
Added Publisher.cs

deakgeek committed:
Updated code snippets with requested changes

arynhuck committed:
A few more fixes that were requested

Almost there! These are only changes that have to do with the Loading
Data tutorial related files.

arynhuck committed:
Changes to querying READMEs

Last of the change requests (for now)

arynhuck commented:
Made the changes! Are we almost good to :shipit: ?

arynhuck committed:
Content fixes. Single command should be last

Moved the `.Single` commands to the end of the query calls.

rdahlman3 commented:
@arynhuck, I'll run through all of the examples one more time to make sure they all work, and then we should be g2g. Thanks for your hard work! :)

rdahlman3 committed:
unnecessary import in librarycontext.cs

rdahlman3 commented on content/efcore/getting-started/samples/loading-related-data-tutorial/multiple-include-methods/Program.cs:
```diff
@@ -0,0 +1,22 @@
+
+            foreach (Edition edition in book.Editions)
+            {
+                Console.WriteLine("{0} - {1} {2}", edition.Book.Title, book.Author.FirstName, book.Author.LastName);
```
This works fine, but doesn't really display the multiple include functionality, because we just print out the book and author like we do in the regular include example. We should print off the editions also to show that we actually do include more than the previous example.

rdahlman3 replied:
I guess we do access the information through a different route which shows the editions are loaded. Maybe we could add a comment to say like - hey, it's printing the same info, but we can see that the editions are also loaded because we're using them to show the same stuff?

deakgeek replied:
Hmm, that's a good point. I think either a comment or printing off something Edition specific, like the Edition year would work. I'm fine with either.

rdahlman3 replied:
Printed off edition year

rdahlman3 committed:
fix indenting in include-method snippet

rdahlman3 committed:
indenting in then-include-method-chain

rdahlman3 committed:
indenting in include-and-then-include-method-chain

rdahlman3 commented on content/efcore/getting-started/loading-related-data.md:
```diff
@@ -0,0 +1,153 @@
+ 
+## Single Layer Inclusion 
+ 
+When loading an entity from the database, EF Core will automatically include properties that are primitive types, but non-primitive types (such as other entities) won't be included. For example, if you run the following code snippet, you will see that the Author field is null: 
```
So, author field is null, so if they run the code snippet, instead of "showing that it is null", it throws an error because we try and print `book.Author.FirstName` and `book.Author.LastName`. We need to either change the wording in the tutorial, or change how we're printing right now.

arynhuck replied:
In the other tutorials, I think they allow errors to be thrown (cuz the wording says "Hey, this results in an error! But try it anyways!").

rdahlman3 replied:
That sounds good to me! We can probably just add a comment into the actual code sample that says something like "Because the `Author` field is null, when we try and print out the `Author.FirstName` and `Author.LastName`, an error will be thrown"

arynhuck replied:
Oh oh oh. The other tutorials also comment out the error code so users have to uncomment it to see

rdahlman3 replied:
Uncommented, put comment above to explain

rdahlman3 committed:
remove unnecessary import in field-not-included Program.cs

rdahlman3 committed:
add code example names so markdown engine can retrieve sample code3

deakgeek committed:
Updated readme to match code snippets

rdahlman3 committed:
fix issues with some snippets

rdahlman3 committed:
Comment out line that will throw error in field-not-included, print out edition year too in multiple-include-methods

rdahlman3 commented:
:shipit:

helmoski committed:
Renames samples subfolders to match chapter keys

helmoski committed:
Switches to printing json for field-not-included example in loading-related-data

helmoski commented:
Explicit Loading Examples don't work, but it is due to the EF Core version in REPL service, not this branch. Bug filed: https://github.com/dotnet/replservice/issues/24

helmoski committed:
Adds inline code snippets where they were missing

helmoski committed:
Pretty prints json

helmoski committed:
Changes from 'Id' to 'BookId'

helmoski committed:
Bolds note

rdahlman3 committed:
move note about needing version 1.1.0
